src/app.ts

```diff
@@ -55,27 +54,100 @@ app.use(cors());
+export type RahaApiContext<
+  Authenticated extends boolean
+> = Authenticated extends true ? AuthenticatedContext : UnauthenticatedContext;
+
+/**
+ * The location of an API endpoint. Uri may contain wildcards that must be
+ * resolved, and only represents a path without a domain to send the request to.
+ */
+interface RouteHandler<Location extends ApiLocation> {
+  location: Location;
+  handler: ReturnType<typeof createApiRoute>;
+}
+
+/**
+ * List of API endpoints the server can respond to.
+ *
+ * If you add a new route or modify an existing one's signature, edit it here.
+ * Also ensure you add routes to the ApiEndpoint type in
+ * routes/ApiEndpoint/index.ts.
+ */
+const apiRoutes: Array<RouteHandler<ApiLocation>> = [
+  {
+    location: operationsRoutes.listOperationsApiLocation,
+    handler: operationsRoutes.listOperations(operationsCollection)
+  },
+  {
+    location: membersRoutes.trustMemberApiLocation,
+    handler: membersRoutes.trust(membersCollection, operationsCollection)
+  },
+  {
+    location: membersRoutes.requestInviteApiLocation,
+    handler: membersRoutes.requestInvite(
+      config,
+      storage,
+      coconutApiKey,
+      membersCollection,
+      operationsCollection
+    )
+  },
+  {
+    location: membersRoutes.giveApiLocation,
+    handler: membersRoutes.give(db, membersCollection, operationsCollection)
+  },
+  {
+    location: meRoutes.sendInviteApiLocation,
+    handler: meRoutes.sendInvite(config, sgMail, membersCollection)
+  },
+  {
+    location: meRoutes.mintApiLocation,
+    handler: meRoutes.mint(db, membersCollection, operationsCollection)
   }
-  ctx.state.toMember = uidDoc;
-  return next();
+];
+
+function createRouter(routes: Array<RouteHandler<ApiLocation>>): Router {
+  return routes.reduce((router, route) => {
+    const { handler, location } = route;
+    const { uri, method } = location;
+    const fullUri = path.join("/api/", uri);
+    switch (method as HttpVerb) {
+      case HttpVerb.GET:
+        return router.get(fullUri, handler);
+      case HttpVerb.POST:
+        return router.post(fullUri, handler);
+      case HttpVerb.PUT:
+        return router.put(fullUri, handler);
+      case HttpVerb.PATCH:
+        return router.patch(fullUri, handler);
+      case HttpVerb.DELETE:
+        return router.delete(fullUri, handler);
+      default:
+        // should be unreachable
+        throw new Error("Invalid HTTP verb");
+    }
+  }, new Router());
 }
 
-const publicRouter = new Router()
-  .param("uid", validateUid)
-  .get("/api/operations", operationsRoutes.index(operations))
-  /* These endpoints are consumed by coconut. */
+const publicRouter = createRouter(
+  apiRoutes.filter(r => !r.location.authenticated)
+)
+  // Coconut video encoding endpoints are not in apiRoutes nor ApiEndpoint for
+  // now. Note: these routes explicitly prefix /api/, unlike the rest of them.
   .post(
-    "/api/members/:uid/notify_video_encoded",
+    "/api/members/:memberId/notify_video_encoded",
```

Comment:
P1 - To me, `memberId` means something like `jane.doe.5555`. Seems name is a bit ambiguous ... how about we transition to semantics like `memberHandle` and `docUid` for the firestore generated document uids?

And using those semantics, plan is to use `docUid` as much as possible so easy for people to edit their `memberHandle` if we enable that?

Reply:
I don’t like uid since it’s just firebase terminology. We want an id for a member, so memberId makes sense; if anything mid should be memberUsername or memberDisplayName, and the whole idea of treating it like an id shouldn’t exist in any of our code

Reply:
I don’t like uid since it’s just firebase terminology. We want an id for a member, so memberId makes sense; if anything mid should be memberUsername or memberDisplayName, and the whole idea of treating it like an id shouldn’t exist in any of our code

Reply:
We should have a single canonical id for members and for anything else really. Right now they’re generated by firebase via the uid field but that’s just an implementation detail and we can very well use any other means of generating ids, in long term this and all other endpoint parameters shouldn’t be tied to firebase since if we actually do decentralize then having everything be in the parlance of a centralized firebase server if anything is even more confusing

Reply:
What would make more sense longer term is to keep firebase uid as its own field and just for now have id === uid; what makes even more sense longer term is to generate our own ids so that id !== uid, and the uid field becomes explicitly tied to firebase rather than serving both the purpose of our own id as well as a binding to firebase’s apis